lib/app/modules/smart/token_refresh_sequence.rb

```diff
@@ -28,133 +31,150 @@ class TokenRefreshSequence < SequenceBase
       end
 
-      test 'Server successfully exchanges refresh token at OAuth token endpoint.' do
+      test 'Refresh token exchange fails when provided invalid Refresh Token.' do
         metadata do
-          id '02'
+          id '01'
           link 'https://tools.ietf.org/html/rfc6749'
           desc %(
-            Server successfully exchanges refresh token at OAuth token endpoint.
-          )
-        end
-
-        oauth2_params = {
-          'grant_type' => 'refresh_token',
-          'refresh_token' => @instance.refresh_token
-        }
-        oauth2_headers = { 'Content-Type' => 'application/x-www-form-urlencoded' }
-
-        if @instance.confidential_client
-          oauth2_headers['Authorization'] = "Basic #{Base64.strict_encode64(@instance.client_id +
-                                                                                ':' +
-                                                                                @instance.client_secret)}"
-        else
-          oauth2_params['client_id'] = @instance.client_id
+            If the request failed verification or is invalid, the authorization server returns an error response.          )
         end
 
-        @token_response = LoggedRestClient.post(@instance.oauth_token_endpoint, oauth2_params, oauth2_headers)
-        assert_response_ok(@token_response)
+        token_response = perform_refresh_request(@instance.client_id, INVALID_REFRESH_TOKEN)
+        assert_response_bad_or_unauthorized token_response
       end
 
-      test 'Data returned from refresh token exchange contains required information encoded in JSON.' do
+      test 'Refresh token exchange fails when provided invalid Client ID.' do
         metadata do
-          id '03'
-          link 'http://www.hl7.org/fhir/smart-app-launch/'
+          id '02'
+          link 'https://tools.ietf.org/html/rfc6749'
           desc %(
-           The EHR authorization server SHALL return a JSON structure that includes an access token or a message indicating that the authorization request has been denied.
-           access_token, token_type, and scope are required. access_token must be Bearer.
-          )
+            If the request failed verification or is invalid, the authorization server returns an error response.          )
         end
 
-        @token_response_headers = @token_response.headers
-        assert_valid_json(@token_response.body)
-        @token_response_body = JSON.parse(@token_response.body)
+        token_response = perform_refresh_request(INVALID_CLIENT_ID, @instance.refresh_token)
+        assert_response_bad_or_unauthorized token_response
+      end
 
-        assert @token_response_body.key?('access_token'), 'Token response did not contain access_token as required'
+      def validate_and_save_refresh_response(token_response)
+        assert_response_ok(token_response)
+        assert_valid_json(token_response.body)
+        token_response_body = JSON.parse(token_response.body)
```

Comment:
Should it?  Yeah, it is a bit wordy to assert then parse instead of doing it in one step, but having an `assert_*` provide more functionality than making an assertion seems strange to me.

Reply:
This is the assertion code:
```
    def assert_valid_json(json)
      JSON.parse(json)
    rescue JSON::ParserError
      raise AssertionException, 'Invalid JSON'
    end
```
I think validating json is always annoying because you validate it by parsing it, so any code like `data = parse(json) if valid(json)` is really just parsing the json twice. But if you'd prefer to treat the assertion as just an assertion rather a "parse the json but raise a special error if it's invalid" method, and leave the code as is that's fine with me.

Reply:
Yeah, understood, i could go either way.  I'm just going to leave this as-is.  Thanks for pointing this out though.